// gatsby-plugin-jss/gatsby-ssr.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _react = _interopRequireDefault(require("react"));

var _reactJss = require("react-jss");

/**
 * Keep track of SheetRegistry for each page
 */
var sheetsRegistryManager = new Map(); // eslint-disable-next-line react/prop-types,react/display-name

exports.wrapRootElement = function (_ref, _ref2) {
  var element = _ref.element,
      pathname = _ref.pathname;
  var _ref2$theme = _ref2.theme,
      theme = _ref2$theme === void 0 ? {} : _ref2$theme;
  var sheets = new _reactJss.SheetsRegistry();
  sheetsRegistryManager.set(pathname, sheets);
  return _react.default.createElement(_reactJss.JssProvider, {
    registry: sheets
  }, _react.default.createElement(_reactJss.ThemeProvider, {
    theme: theme
  }, element));
};

exports.onRenderBody = function (_ref3) {
  var setHeadComponents = _ref3.setHeadComponents,
      pathname = _ref3.pathname;
  var sheets = sheetsRegistryManager.get(pathname);

  if (sheets) {
    setHeadComponents([_react.default.createElement("style", {
      type: "text/css",
      id: "server-side-jss",
      key: "server-side-jss",
      dangerouslySetInnerHTML: {
        __html: sheets.toString()
      }
    })]);
    sheetsRegistryManager.delete(pathname);
  }
};
